src/ModelContextProtocol.AspNetCore/AuthorizationFilterSetup.cs

```diff
@@ -0,0 +1,360 @@
+            }
+
+            return result;
+        });
+    }
+
+    private void ConfigureGetPromptFilter(McpServerOptions options)
+    {
+        options.Filters.GetPromptFilters.Add(next => async (context, cancellationToken) =>
+        {
+            context.Items[AuthorizationFilterInvokedKey] = true;
+
+            var authResult = await GetAuthorizationResultAsync(context.User, context.MatchedPrimitive, context.Services, context);
+            if (!authResult.Succeeded)
+            {
+                throw new McpException("Access forbidden: This prompt requires authorization.", McpErrorCode.InvalidRequest);
+            }
+
+            return await next(context, cancellationToken);
+        });
+    }
+
+    private void CheckGetPromptFilter(McpServerOptions options)
+    {
+        options.Filters.GetPromptFilters.Add(next => async (context, cancellationToken) =>
+        {
+            if (HasAuthorizationMetadata(context.MatchedPrimitive)
+                && !context.Items.ContainsKey(AuthorizationFilterInvokedKey))
+            {
+                throw new InvalidOperationException("Authorization filter was not invoked for prompts/get operation, but authorization metadata was found on the prompt. Ensure that AddAuthorizationFilters() is called on the IMcpServerBuilder to configure authorization filters.");
+            }
+
+            return await next(context, cancellationToken);
+        });
+    }
+
+    /// <summary>
+    /// Filters a collection of items based on authorization policies in their metadata.
+    /// For list operations where we need to filter results by authorization.
+    /// </summary>
+    private async ValueTask FilterAuthorizedItemsAsync<T>(IList<T> items, Func<T, IMcpServerPrimitive?> primitiveSelector,
+        ClaimsPrincipal? user, IServiceProvider? requestServices, object context)
+    {
+        for (int i = items.Count - 1; i >= 0; i--)
+        {
+            var authorizationResult = await GetAuthorizationResultAsync(
+                user, primitiveSelector(items[i]), requestServices, context);
+
+            if (!authorizationResult.Succeeded)
+            {
+                items.RemoveAt(i);
+            }
+        }
+    }
+
+    private async ValueTask<AuthorizationResult> GetAuthorizationResultAsync(
+        ClaimsPrincipal? user, IMcpServerPrimitive? primitive, IServiceProvider? requestServices, object context)
+    {
+        if (!HasAuthorizationMetadata(primitive))
+        {
+            return AuthorizationResult.Success();
+        }
+
+        if (policyProvider is null)
+        {
+            throw new InvalidOperationException($"You must call AddAuthorization() because an authorization related attribute was found on {primitive.Id}");
+        }
+
+        // TODO: Cache policy lookup. We would probably use a singleton (not-static) ConditionalWeakTable<IMcpServerPrimitive, AuthorizationPolicy?>.
```

Comment:
TODO for when?

Reply:
I filed the issue at #762, and I'll remove this TODO. `IAuthorizationPolicyProvider.AllowsCachingPolicies` was only added in .NET 7, and things worked pretty well before then, so I don't think this needs to be a huge priority.